### ISSUE TITLE
Absolutely positioned block box are collaping margin with its previous sibling

### ISSUE THREAD

**Issue by shinglyu** [servo/servo#12676]
According to the [spec](https://drafts.csswg.org/css-position-3/#position-property), a block element with `position:absolute` and NO `top`, `bottom`, `left`, `right` defined should NOT collapse with any margin. 

> Though absolutely positioned boxes may have margins, those margins do not collapse with any other margins. 

But the following test case shows that that red box is collapsing with the `#margin-box`'s bottom margin.

```
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html>
    <head>
        <style type="text/css">
            #margin-box {
                background-color:blue;
                width: 100px;
                height: 10px;
                margin-bottom: 50px;
            }

            #reference-overlapped-red {
                position: absolute;
                background-color: red;
                width: 100px;
                height: 100px;
                z-index:-1;
            }

            #test-overlapping-green {
                background-color: green;
                width: 100px;
                height: 100px;
            }
        </style>
    </head>
    <body>
        <div id="margin-box"></div>
        <div id="reference-overlapped-red"></div>
        <div id="test-overlapping-green"></div>
    </body>
</html>
```
Screenshot: (left=servo, right=gecko)
![selection_015](https://cloud.githubusercontent.com/assets/3250983/17289169/fda239fe-580a-11e6-888d-0dec01edd30c.png)

**Comment by shinglyu** [comment]
This blocks https://github.com/servo/servo/issues/12610 in some sense because all the `tests/wpt/css-tests/css-flexbox-1_dev/html/flex-minimum-width-flex-items-00*.html` tests is broken because of this.

**Comment by shinglyu** [comment]
I've already identify the problem, but my naive patch caused some regression. Need to Double check with the spec to fix the regression.

**Comment by shinglyu** [comment]
The problem is this comment in `block.rs`:

```
                    // Assume that the *hypothetical box* for an absolute flow starts immediately
                    // after the block-end border edge of the previous flow.
```

The hypothetical box has to consider the previous flow's bottom margin if it has no `top`, `bottom`, `left`, `right` property.

**Comment by shinglyu** [comment]
I think I fixed it. My patch fixed 99 reftests but made one fail: `/css21_dev/html4/left-offset-position-fixed-001.htm`. 

I found out that the test was a false positive, both the test and reference are rendered incorrectly, so the test passes, not the reference are fixed by the patch, but the test file failed for other reason. I'll check weather the bug is related, otherwise I should land my patch and create another bug for the `left-offset-position-fixed-001.htm` case

This is a proof that the reference file was rendered incorrectly:
![selection_027](https://cloud.githubusercontent.com/assets/3250983/17552248/68a9a4a8-5f31-11e6-8bb3-d6e2da482665.png)
